lib/vec.dl

```diff
@@ -0,0 +1,152 @@
+ * resulting vectors, returning a flat vector. */
+function flatmap(v: Vec<'A>, f: function('A): Vec<'B>): Vec<'B> {
+    var res = vec_empty();
+    for (x in v) {
+        res.append(f(x))
+    };
+    res
+}
+
+/* Sorts the vector with a key extraction function, but may not preserve the
+ * order of equal elements. */
+function sort_by(v: mut Vec<'A>, f: function('A): 'B) {
+    vec_sort_by(v, f)
+}
+
+/* Returns the element that gives the minimum value from the specified function.
+ * If several elements are equally minimum, the first element is returned.
+ * If the vector is empty, `None` is returned. */
+function min_by(v: Vec<'A>, f: function('A): 'B): Option<'A> {
+    vec_min_by(v, f)
+}
+
+/* Returns the element that gives the maximum value from the specified function.
+ * If several elements are equally maximum, the first element is returned.
+ * If the vector is empty, `None` is returned. */
+function max_by(v: Vec<'A>, f: function('A): 'B): Option<'A> {
+    vec_max_by(v, f)
+}
+
+/* Returns the first element of the vector that satisfies predicate `f` or
+ * `None` if none of the elements satisfy the predicate. */
+function find(v: Vec<'A>, f: function('A): bool): Option<'A> {
+    for (x in v) {
+        if (f(x)) {
+            return Some{x}
+        }
+    };
+    None
+}
+
+/* Returns a vector containing only those elements in `v` that satisfy predicate `f`,
+ * preserving the order of the elements in the original vector. */
+function filter(v: Vec<'A>, f: function('A): bool): Vec<'A> {
+    var res = vec_empty();
+    for (x in v) {
+        if (f(x)) {
+            res.push(x)
+        }
+    };
+    res
+}
+
+/* Both filters and maps the vector.
+ *
+ * Calls the closure on each element of the vector.  If the closure returns
+ * `Some{element}`, then that element is returned. */
+function filter_map(v: Vec<'A>, f: function('A): Option<'B>): Vec<'B> {
+    var res = vec_empty();
+    for (x in v) {
+        match (f(x)) {
+            None -> (),
+            Some{y} -> res.push(y)
+        }
+    };
+    res
+}
+
+
+/* Retains only the elements specified by the predicate.
+ * In other words, remove all elements `e` such that `f(e)` returns `false`.
+ * This method operates in place, visiting each element exactly once in the
+ * original order, and preserves the order of the retained elements. */
+function retain(v: mut Vec<'A>, f: function('A): bool) {
+    var del: s64 = 0;
+    var len = v.len();
+
+    for (i in range(0, (len as isize) - 1, 1)) {
+        var x = v.nth(i as usize).unwrap_or_default();
+        if (not f(x)) {
+            del = del + 1
+        } else if (del > 0) {
+            v.update_nth((i - del) as usize, x);
+        }
+    };
+    if (del > 0) {
+        v.truncate(len - (del as usize));
+    }
+}
+
+/* Returns `true` iff all elements of the vector satisfy predicate `f`. */
+function all(v: Vec<'A>, f: function('A): bool): bool {
+    for (x in v) {
```

Comment:
in the future we'll just have an Iterable trait...

Reply:
Speaking of which, we don't have a way to implement lazy iterators at the moment (as they require either lifetimes or reference counting), so for example, if we have  `e.filter(..).all(..)`, then `filter` will iterate to the end of the collection even if `all` can terminate early.

Reply:
I think it should be possible to do that, but I don't know enough about Rust to see the potential difficulties.
Each element should have a reference in exactly one iterator, so I imagine it should be possible to do that.

Reply:
Example: `v.map(...).find(..)`  

The return value of `map` is `Vec`.  DDlog returns by-value, so we need to create a fully populated vector before passing it to the next transformer (`find`).

Let's say, we change `map` to return `Iterator<>` instead. The iterator object would internally hold a reference to the original vector `v`.  This is safe as long as the iterator is destroyed before anyone tries to modify or destroy `v`, which is the case here. But not in this code:

```
var iter = v.map();
v.push(x);
iter.find(...);
```

In Rust this cannot happen thanks to the borrow checker.  In fact, the above code will be rejected by the Rust compiler.  But DDlog does not currently have a way to track ownership.

Reply:
perhaps you should be able to obtain an iterator only from an immutable vector. this is a reasonable constraint.

Reply:
We'll need to add a concept of mutable variables  for that, but this still does not address the lifetime issue.  The iterator can outlive the vector it was create from. Without reference counting or a borrow checker we cannot support this safely.

Reply:
how about an iterator that actually consumes the vector?

Reply:
That might work, but again this is a pretty big change to the language (we'd have to annotate functions to consume their arguments and enforce that they are not used afterwards). And then you don't always want to consume the argument.  Point is, you cannot easily retrofit a new ownership model in the language.  There is a lot of complexity in Rust that takes care of that.